docs/source/whatsnew/version3_widget_migration.rst

```diff
@@ -0,0 +1,318 @@
+   instead of N times where N is the number of displayed views.
+
+Python
+~~~~~~
+
+Generally, custom widget Python code can remain unchanged. If you
+distribute your custom widget, you may be using ``display`` and
+``Javascript`` to publish the widget's Javascript to the front-end. That
+is no longer the recommended way of distributing widget Javascript.
+Instead have the user install the Javascript to his/her nbextension
+directory or their profile's static directory. Then use the new
+``_view_module`` and ``_model_module`` traitlets in combination with
+``_view_name`` and ``_model_name`` to instruct require.js on how to load
+the widget's Javascript. The Javascript is then loaded when the widget
+is used for the first time.
+
+Details
+-------
+
+Asynchronous
+~~~~~~~~~~~~
+
+In the IPython 2.x series the only way to register custom widget views
+and models was to use the registry in the widget manager. Unfortunately,
+using this method made packing custom widgets difficult. The widget
+maintainer had to either use the rich display framework to push the
+widget's Javascript to the notebook or instruct the users to install the
+Javascript by hand in a custom profile. With the first method, the
+maintainer would have to be careful about when the Javascript was pushed
+to the font-end. If the Javascript was pushed on Python widget
+``import``, the widgets wouldn't work after page refresh. This is
+because refreshing the page does not restart the kernel, and the Python
+``import`` statement only runs once in a given kernel instance (unless
+you reload the Python modules, which isn't straight forward). This meant
+the maintainer would have to have a separate ``push_js()`` method that
+the user would have to call after importing the widget's Python code.
+
+Our solution was to add support for loading widget views and models
+using require.js paths. Thus the comm and widget frameworks now support
+lazy loading. To do so, everything had to be converted to asynchronous
+code. HTML5 promises are used to accomplish that
+(`#6818 <https://github.com/ipython/ipython/pull/6818>`__,
+`#6914 <https://github.com/ipython/ipython/pull/6914>`__).
+
+Symmetry
+~~~~~~~~
+
+In IPython 3.0, widgets can be instantiated from the front-end
+(`#6664 <https://github.com/ipython/ipython/pull/6664>`__). On top of
+this, a widget persistence API was added
+(`#7163 <https://github.com/ipython/ipython/pull/7163>`__,
+`#7227 <https://github.com/ipython/ipython/pull/7227>`__). With the
+widget persistence API, you can persist your widget instances using
+Javascript. This makes it easy to persist your widgets to your notebook
+document (with a small amount of custom JS). By default, the widgets are
+persisted to your web browsers local storage which makes them reappear
+when your refresh the page.
+
+Smaller Changes
+~~~~~~~~~~~~~~~
+
+-  Latex math is supported in widget ``description``\ s
+   (`#5937 <https://github.com/ipython/ipython/pull/5937>`__).
```

Comment:
For future reference, we have an extension so you can write these like:

```
:ghissue:`5937`
```

But it's not necessary to go through and replace all these.